File: app/src/main/java/com/blackmirror/dongda/presenter/DetailInfoPresenter.java
package com.blackmirror.dongda.presenter;

import com.blackmirror.dongda.domain.Interactor.DetailInfoUseCase;
import com.blackmirror.dongda.domain.Interactor.LikePopUseCase;
import com.blackmirror.dongda.domain.Interactor.LikePushUseCase;
import com.blackmirror.dongda.domain.model.BaseDataBean;
import com.blackmirror.dongda.domain.model.DetailInfoDomainBean;
import com.blackmirror.dongda.domain.model.LikePopDomainBean;
import com.blackmirror.dongda.domain.model.LikePushDomainBean;
import com.blackmirror.dongda.ui.Contract;
import com.blackmirror.dongda.utils.AppConstant;
import com.blackmirror.dongda.utils.LogUtils;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Create By Ruge at 2018-05-18
 */
public class DetailInfoPresenter implements Contract.DetailInfoPresenter {

    @Inject
    DetailInfoUseCase detailInfoUseCase;

    @Inject
    LikePushUseCase pushUseCase;

    @Inject
    LikePopUseCase popUseCase;

    private Contract.DetailInfoView detailInfoView;

    @Inject
    public DetailInfoPresenter(Contract.DetailInfoView detailInfoView) {
        this.detailInfoView = detailInfoView;
    }

    @Override
    public void getDetailInfo(String service_id) {
        detailInfoUseCase.getDetailInfo(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<DetailInfoDomainBean>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(DetailInfoDomainBean bean) {
                        if (detailInfoView == null){
                            return;
                        }
                        if (bean.isSuccess){
                            detailInfoView.onGetDetailInfoSuccess(bean);
                        }else {
                            detailInfoView.onGetDataError(bean);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        LogUtils.e(DetailInfoPresenter.class,e);
                        if (detailInfoView == null){
                            return;
                        }
                        detailInfoView.onGetDataError(getErrorData(e));

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void likePush(String service_id) {
        pushUseCase.likePush(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<LikePushDomainBean>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(LikePushDomainBean bean) {
                        if (detailInfoView == null){
                            return;
                        }
                        if (bean.isSuccess){
                            detailInfoView.onLikePushSuccess(bean);
                        }else {
                            detailInfoView.onGetDataError(bean);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        LogUtils.e(DetailInfoPresenter.class,e);
                        if (detailInfoView == null){
                            return;
                        }
                        detailInfoView.onGetDataError(getErrorData(e));

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void likePop(String service_id) {
        popUseCase.likePop(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<LikePopDomainBean>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(LikePopDomainBean bean) {
                        if (detailInfoView == null){
                            return;
                        }
                        if (bean.isSuccess){
                            detailInfoView.onLikePopSuccess(bean);
                        }else {
                            detailInfoView.onGetDataError(bean);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        LogUtils.e(DetailInfoPresenter.class,e);
                        if (detailInfoView == null){
                            return;
                        }
                        detailInfoView.onGetDataError(getErrorData(e));

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    private BaseDataBean getErrorData(Throwable e) {
        BaseDataBean bean = new BaseDataBean();
        bean.code = AppConstant.NET_UNKNOWN_ERROR;
        bean.message = e.getMessage();
        return bean;
    }
}

File: app/src/main/java/com/blackmirror/dongda/presenter/DetailInfoPresenter.kt
package com.blackmirror.dongda.presenter

import com.blackmirror.dongda.kdomain.interactor.getDetailInfoImpl
import com.blackmirror.dongda.kdomain.interactor.likePopImpl
import com.blackmirror.dongda.kdomain.interactor.likePushImpl
import com.blackmirror.dongda.kdomain.model.BaseDataBean
import com.blackmirror.dongda.ui.Contract
import com.blackmirror.dongda.utils.AppConstant
import com.blackmirror.dongda.utils.logE
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Create By Ruge at 2018-05-18
 */
class DetailInfoPresenter @Inject constructor(val detailInfoView: Contract.DetailInfoView?) : Contract.DetailInfoPresenter {

    override fun getDetailInfo(service_id: String) {
        getDetailInfoImpl(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.isSuccess) {
                        detailInfoView?.onGetDetailInfoSuccess(it)
                    } else {
                        detailInfoView?.onGetDataError(it)
                    }
                }, {
                    logE(message = DetailInfoPresenter::class.java.simpleName, exception = it)

                    detailInfoView?.onGetDataError(getErrorData(it))
                })
    }

    override fun likePush(service_id: String) {
        likePushImpl(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.isSuccess) {
                        detailInfoView?.onLikePushSuccess(it)
                    } else {
                        detailInfoView?.onGetDataError(it)
                    }
                }, {
                    logE(message = DetailInfoPresenter::class.java.simpleName, exception = it)
                    detailInfoView?.onGetDataError(getErrorData(it))
                })
    }

    override fun likePop(service_id: String) {
        likePopImpl(service_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.isSuccess) {
                        detailInfoView?.onLikePopSuccess(it)
                    } else {
                        detailInfoView?.onGetDataError(it)
                    }
                }, {
                    logE(message = DetailInfoPresenter::class.java.simpleName, exception = it)
                    detailInfoView?.onGetDataError(getErrorData(it))
                })
    }

    private fun getErrorData(e: Throwable): BaseDataBean {
        val bean = BaseDataBean()
        bean.code = AppConstant.NET_UNKNOWN_ERROR
        bean.message = e.message
        return bean
    }
}
